spinalcordtoolbox/qmri/mt.py

```diff
@@ -44,8 +44,13 @@ def compute_mtr(nii_mt1, nii_mt0, threshold_mtr=100):
     :param threshold_mtr: float: value above which number will be clipped
     :return: nii_mtr
     """
+    # Convert input before compute MTR
+    nii_mt1.change_type('float32')
+    nii_mt0.change_type('float32')
```

Comment:
While this may be an appropriate fix, I'm also curious to find out what the root cause of the problem actually is!

((For example, is there an issue with our `divide_after_removing_zero` function? I'd love to find out the place where the errors get introduced at... :thinking:))

So, I think it could be a good idea to create some int16 data (as you mentioned), then explore the behavior of the functions to determine the place where the differences get introduced...

Reply:
It's probably simplest to just open the existing `mt.nii.gz` files using SCT's [`Image`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/e106992c42ac8be6e74e9bf8cba53a9a71b42dd6/spinalcordtoolbox/image.py#L248) class, convert them, then re-save them to a temporary directory using pytest's [`tmp_path`](https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture)?

Reply:
For sure! I tried to see things easily to begin with but i never say no to some analysis ;)
I will do the investigation and report on this PR as soon as I have something.

@joshuacwnewton Do you have some idea how i can have some others file of tests ? When i do it with `sct_testing_data/mt/mt0.json` and `sct_testing_data/mt/mt1.json` it just say 0 voxels for both int16 and float32 input...

@jcohenadad Maybe the file you got from the issue's author are public and we can add them in the testing base ? I'm not aware of the intellectual property of this kind of file.

Reply:
Ahh, so. The json files are different than the image files. The files you want end in the `.nii.gz` extension. (`.nii` refers to the [NIfTI file format](https://radiopaedia.org/articles/nifti-file-format), and `.gz` is additional compression to save space.)

![Screenshot from 2022-01-10 14-03-15](https://user-images.githubusercontent.com/16181459/148824019-c10cbd15-e0bf-44cd-9037-cc7189577585.png)

Reply:
@joshuacwnewton Yes, I got that and did some personal time research on NifTI & DICOM. I also played a bit with FSLEyes to make sure I was able to read them. 

But actually when I do:

```
❯ sct_compute_mtr -mt0 sct_testing_data/mt/mt0.nii.gz -mt1 sct_testing_data/mt/mtr.nii.gz
```

I got:
```
--
Spinal Cord Toolbox (git-enhancement-help_linux_user_to_find_wsl_section-ea2ad7aaa76e2e1c8a9f6ba92762fa9ee3329cb0*)

sct_compute_mtr -mt0 sct_testing_data/mt/mt0.nii.gz -mt1 sct_testing_data/mt/mtr.nii.gz
--


Compute MTR...
Found 0 voxels with value=0. These will be replaced by nan.
Threshold to clip values: +/- 100
File ./mtr.nii.gz already exists. Will overwrite it.

Done! To view results, type:
fsleyes sct_testing_data/mt/mt0.nii.gz sct_testing_data/mt/mtr.nii.gz ./mtr.nii.gz &
```

But it say `Found 0 voxels`, and I guess that it's not good since the issue say that a lot of voxels have wrong values. 

---

The two mains problems I have are:

 - Understanding of the validation process: I do not know how to validate if the output files are correct or not
 - Miss knowledge on input files: As you say, there are some `.json` files also, but i'm not sure what they are. From what I learn on IRM lately, i guess they are the IRM process configuration to take images (ie: echo time, repetition time, etc..)

Reply:
Oh, sorry! I thought you were passing `.json` files to `sct_compute_mtr`. I misunderstood. :)

> ```
> Found 0 voxels with value=0. These will be replaced by nan.
> ```

This seems like a bug in the function to me! The function seems like it's working properly, so I feel like this log message should maybe be skipped to prevent confusion, in the case where there are no voxels that need to be replaced. 

> Understanding of the validation process: I do not know how to validate if the output files are correct or not

In `sct_testing_data`, there is already a pre-computed [`mtr.nii.gz`](https://github.com/spinalcordtoolbox/sct_testing_data/blob/master/mt/mtr.nii.gz) image, too. So, one suggestion I have is to treat the existing image like "ground truth", then output the newly calculated image to a different folder (temporary directory), and compare the ground truth to the newly calculated file, and make sure that the two files are within a specific threshold.

Similar to other tests we have? For example: 

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/0bef2991f0c05251fbf43df8d1fc0360960565c2/testing/cli/test_cli_sct_analyze_texture.py#L19-L22